cocos/animation/core/pose-allocator.ts

```diff
@@ -0,0 +1,156 @@
+import { assertIsTrue } from '../../core/data/utils/asserts';
+import { Pose } from './pose';
+import { TransformArray } from './transform-array';
+
+const MAX_POSE_PER_PAGE = 8;
+
+const allocationInfoTag = Symbol('PoseAllocator');
+
+type PagedPose = Pose & {
+    [allocationInfoTag]: AllocationInfo;
+};
+
+function isPagedPose (pose: Pose): pose is PagedPose {
+    return allocationInfoTag in pose;
+}
+
+export class PoseAllocator {
+    constructor (transformCount: number, metaValueCount: number) {
+        this._transformCount = transformCount;
+        this._metaValueCount = metaValueCount;
+    }
+
+    public get allocatedCount () {
+        return this._allocatedCount;
+    }
+
+    public allocatePose (): Pose {
+        ++this._allocatedCount;
+        const { _pages: pages } = this;
+        const nPages = pages.length;
+        for (let iPage = 0; iPage < nPages; ++iPage) {
```

Comment:
What is the frequency of allocation during the run? If frequency is high, you may need to design the `freeList` here to avoid iterating through all pages

Reply:
The allocation happens frequencely, for example:
- Every non-additive clip motion causes 1 allocation.
- Every additive clip motion causes 2 allocation.
- Every animation blend in theory causes 1 allocation, but I optimize it to no allocation.

Reply:
> The allocation happens frequencely, for example:
Every non-additive clip motion causes 1 allocation.
Every additive clip motion causes 2 allocation.
Every animation blend in theory causes 1 allocation, but I optimize it to no allocation.

Perframe ?

Reply:
Yes per frame